pdns/dnsdistdist/dnsdist-tcp-downstream.cc

```diff
@@ -0,0 +1,543 @@
+        conn->d_responseBuffer.reserve(conn->d_responseSize + /* we will need to prepend the size later */ 2);
+        conn->d_responseBuffer.resize(conn->d_responseSize);
+        conn->d_currentPos = 0;
+      }
+    }
+
+    if (conn->d_state == State::readingResponseFromBackend) {
+      DEBUGLOG("reading response from backend");
+      iostate = tryRead(fd, conn->d_responseBuffer, conn->d_currentPos, conn->d_responseSize - conn->d_currentPos);
+      if (iostate == IOState::Done) {
+        DEBUGLOG("got response from backend");
+        try {
+          iostate = conn->handleResponse(conn, now);
+        }
+        catch (const std::exception& e) {
+          vinfolog("Got an exception while handling TCP response from %s (client is %s): %s", conn->d_ds ? conn->d_ds->getName() : "unknown", conn->d_currentQuery.d_idstate.origRemote.toStringWithPort(), e.what());
+        }
+      }
+    }
+
+    if (conn->d_state != State::idle &&
+        conn->d_state != State::sendingQueryToBackend &&
+        conn->d_state != State::readingResponseSizeFromBackend &&
+        conn->d_state != State::readingResponseFromBackend) {
+      vinfolog("Unexpected state %d in TCPConnectionToBackend::handleIO", static_cast<int>(conn->d_state));
+    }
+  }
+  catch (const std::exception& e) {
+    /* most likely an EOF because the other end closed the connection,
+       but it might also be a real IO error or something else.
+       Let's just drop the connection
+    */
+    vinfolog("Got an exception while handling (%s backend) TCP query from %s: %s", (conn->d_ioState->getState() == IOState::NeedRead ? "reading from" : "writing to"), conn->d_currentQuery.d_idstate.origRemote.toStringWithPort(), e.what());
+    if (conn->d_state == State::sendingQueryToBackend) {
+      ++conn->d_ds->tcpDiedSendingQuery;
+    }
+    else {
+      ++conn->d_ds->tcpDiedReadingResponse;
+    }
+
+    /* don't increase this counter when reusing connections */
+    if (conn->d_fresh) {
+      ++conn->d_downstreamFailures;
+    }
+
+    /* remove this FD from the IO multiplexer */
+    iostate = IOState::Done;
+    connectionDied = true;
+  }
+
+  if (connectionDied) {
+
+    bool reconnected = false;
+    DEBUGLOG("connection died, number of failures is "<<conn->d_downstreamFailures<<", retries is "<<conn->d_ds->retries);
+
+    if ((!conn->d_usedForXFR || conn->d_queries == 0) && conn->d_downstreamFailures < conn->d_ds->retries) {
+
+      conn->d_ioState->reset();
+      ioGuard.release();
+
+      try {
+        if (conn->reconnect()) {
+          conn->d_ioState = make_unique<IOStateHandler>(conn->d_clientConn->getIOMPlexer(), conn->d_socket->getHandle());
+
+          /* we need to resend the queries that were in flight, if any */
+          for (auto& pending : conn->d_pendingResponses) {
+            conn->d_pendingQueries.push_back(std::move(pending.second));
+            if (!conn->d_usedForXFR) {
+              --conn->d_ds->outstanding;
```

Comment:
Maybe a >0 check here to detect accounting errors wrt `outstanding`?

Reply:
We have tried that in the past, and unfortunately it did not really help because it covered issues up for light loads, while the race between the test and the decrement was happening quite frequently at high loads.

Reply:
right. making it safe would be  more complex indeed.